### DISCUSSION TITLE
Idea to improve RF noise from SPI lines

### DISCUSSION THREAD

**Discussion by zygmund2000** [PE5PVB/TEF6686_ESP32#362] [Ideas]
However, the main reason for the idea is the noticeable emission of signals coming from the SPI bus, good choice is to use resistors close to pins outputs, but the best solution is to change speed of rising and falling edges internally in MCU.
ESP32 also has this feature and I found in documentation how to use that:
[https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html#_CPPv425gpio_set_drive_capability10gpio_num_t16gpio_drive_cap_t](url)
Maybe there is someone here who can help me find it in the code, because it can reduce spurs up to 10db without loss of data.

Thanks

**Comment by PE5PVB** [comment]
The link you provided doesn't work.

**Reply by HyperDX** [reply]
Try here:
[https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html#_CPPv425gpio_set_drive_capability10gpio_num_t16gpio_drive_cap_t](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html)

**Reply by zygmund2000** [reply]
Yes, I added to main code .ino two lines:
`gpio_set_drive_capability((gpio_num_t) 18, GPIO_DRIVE_CAP_0);`
`gpio_set_drive_capability((gpio_num_t) 23, GPIO_DRIVE_CAP_0);`

And now signal is much more smooth, no overshots, spurs are lower, and LCD still work without problems.

**Comment by PE5PVB** [comment]
Thanks. I added:
  gpio_set_drive_capability((gpio_num_t) 5, GPIO_DRIVE_CAP_0);
  gpio_set_drive_capability((gpio_num_t) 16, GPIO_DRIVE_CAP_0);
  gpio_set_drive_capability((gpio_num_t) 17, GPIO_DRIVE_CAP_0);
  gpio_set_drive_capability((gpio_num_t) 18, GPIO_DRIVE_CAP_0);
  gpio_set_drive_capability((gpio_num_t) 19, GPIO_DRIVE_CAP_0);
  gpio_set_drive_capability((gpio_num_t) 23, GPIO_DRIVE_CAP_0);
  
  Much more quiet now.

**Comment by kkonradpl** [comment]
The I²C bus interference can be also lowered with:
```
gpio_set_drive_capability((gpio_num_t)21, GPIO_DRIVE_CAP_0);
gpio_set_drive_capability((gpio_num_t)22, GPIO_DRIVE_CAP_0);
```

With this settings the interference level is decreased or is even below noise level. See the example, this is PE5PVB board without antenna connected and without display, running my sketch: https://www.youtube.com/watch?v=GVjCEUiXoLE
Antenna input alignment is used for GPIO_DRIVE_CAP manipulation, i.e.:
0 - GPIO_DRIVE_CAP_0
1 - GPIO_DRIVE_CAP_1
2 - GPIO_DRIVE_CAP_2
3 - GPIO_DRIVE_CAP_3
4 - GPIO_DRIVE_CAP_DEFAULT

**Comment by kkonradpl** [comment]
Just an update: this test above was at 400 kHz bus speed.

On regular 100 kHz there is audible interference every 100 kHz without antenna connected, which sounds like a weak empty carrier. With the GPIO_DRIVE_CAP_0 set, the interference on the FM band is gone.

**Comment by PE5PVB** [comment]
I added the lines, and it works.
